library/src/main/java/com/datatorrent/lib/io/fs/HDFSFileMerger.java

```diff
@@ -0,0 +1,218 @@
+/**
+ * Licensed to the Apache Software Foundation (ASF) under one
+ * or more contributor license agreements.  See the NOTICE file
+ * distributed with this work for additional information
+ * regarding copyright ownership.  The ASF licenses this file
+ * to you under the Apache License, Version 2.0 (the
+ * "License"); you may not use this file except in compliance
+ * with the License.  You may obtain a copy of the License at
+ *
+ *   http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing,
+ * software distributed under the License is distributed on an
+ * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+ * KIND, either express or implied.  See the License for the
+ * specific language governing permissions and limitations
+ * under the License.
+ */
+
+package com.datatorrent.lib.io.fs;
+
+import java.io.IOException;
+
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+import org.apache.hadoop.fs.FileStatus;
+import org.apache.hadoop.fs.FileSystem;
+import org.apache.hadoop.fs.Path;
+
+import com.google.common.annotations.VisibleForTesting;
+
+import com.datatorrent.api.Context.OperatorContext;
+import com.datatorrent.lib.io.fs.Synchronizer.OutputFileMetadata;
+
+/**
+ * HDFS file merger extends file merger to optimize for HDFS file copy usecase.
+ * This uses fast merge from HDFS if destination filesystem is same as
+ * application filesystem.
+ */
+public class HDFSFileMerger extends FileMerger
+{
+  /**
+   * Fast merge is possible if append is allowed for output file system.
+   */
+  private transient boolean fastMergeActive;
```

Comment:
Because fastMergeActive is quick check. FastMergeDecisionMaker.isFastMergePossible is expensive check. We do second check only if quick check is passed.